### PR TITLE
ffmpeg: switch to 4.3-libreelec integration branch

### PR DESCRIPTION
**Pull request by HiassofT**
Trying to deal with ffmpeg feature branches in the LE buildsystem  is a huge PITA.

It starts with having to find the right merge base points to create the patchsets, sometimes having to use diff instead of format-patch as patch/git apply doesn't cope well with non-linear histories and it doesn't end at patch/git apply also failing if some feature branch merged another one (as we currently have eg on RPi4 which also contains the v4l2 request api patches).

git OTOH does a way better job at that, a single octopus merge command (separate merges work, too) is enough to create a working ffmpeg branch from all feature branches and if all involved parties play nice and don't destroy history of already merged branches we can simply fetch + merge updates.

Several of the global ffmpeg patches seem to be leftovers from ancient RPi/MMAL changes that never went anywhere except into LE - so I dropped those.

The only two patches that looked somewhat relevant are ffmpeg-99.1009-dav1d-fix-multithreaded-av1-sw-decoding.patch/"libavcodec/libdav1d: add libdav1d_get_format method to call ff_get_format"  from @lrusak and ffmpeg-99.1010-yuv2rgb-logspam.patch from @chewitt 

As I couldn't find any remote branches for these patches I added them to the 4.3-libreelec-misc branch https://github.com/LibreELEC/FFmpeg/tree/4.3-libreelec-misc - please check if they are still needed and esp for the dav1d patch it would be interesting to know why it's not upstream.

All other patchsets have feature branches which I merged together (including the 4.3-libreelec-misc) into the 4.3-libreelec branch https://github.com/LibreELEC/FFmpeg/tree/4.3-libreelec

Build test for RPi4, RPi2 and RK3328 (these were the build trees I had on my machine) worked fine, but I haven't runtime tested it yet.

See this PR also as an RFC/proposal how we should deal with ffmpeg patches in LE. Merging changes from feature-branches into LE integration branches (4.3-libreelec ATM) seems to be the best option to me, but of course I'm open to other suggestions/ideas/... as well.

### CODE REVIEW

**Comment by HiassofT** [comment]
I forgot the almost most important thing: I included https://github.com/LibreELEC/FFmpeg/blob/4.3-libreelec/LibreELEC.md in the ffmpeg branch which describes where the actual feature branches are located - this really important info has been missing previously with the patchsets in the LE buildsystem

**Comment by Kwiboo** [comment]
Great work to track down all source branches and filtering out all patches no longer needed!

As mentioned on Slack I do not think using an integration branch is a good idea since I believe that makes it harder to understand what has changed and also makes it harder to replace one of the feature branches with a new set of patches for development and/or local testing.

I have a counter proposal in where we continue to use patches but define where they come from and how to generate the patchsets, see https://github.com/Kwiboo/LibreELEC.tv/tree/ffmpeg-4.3/packages/multimedia/ffmpeg/patches
Guess we can also move some of the feature branches into the LE FFmpeg repo to make it clearer where the source of the patchset come from.

Also from what I can tell we only need the v4l2 stateless custom parts for rpi, so we can probably reduce / remove that patch in time.

**Comment by HiassofT** [comment]
Thanks a lot for your feedback and for pointing out that RPi patches are touching core request code - so keeping RPi separate is definitely a good idea (also if kernel and therefore request APi versions diverge).

I think keeping track of the feature branches, maybe moving some of them to the LE ffmpeg repo, is the most important thing - independent from the decision to use an integration branch or not.

I had a rather similar idea as you this morning, keeping the feature branch info in the LE repo, and implemented an integration script to create/update the integration branch and also record merge info (repo, branch, change summary, like linux merge commits have).

If you'd like to keep the patchsets in the LE tree then my preference would be to put that info into the patchset update commits though as that's easily accesible by just doing a git log.

**Comment by HiassofT** [comment]
ATM we're juggling around with `PKG_PATCH_DIRS+=...` in various places in ffmpeg package.mk and also have to take care about configure options that are introduced by these feature branches (like enable/disable v4l2-request).

Leaving RPi out for a moment it seems to me that it should be safe to always pull in v4l2-drmprime and v4l2-request feature branches and just use enable/disable configure options. eg Generic will disable both ATM, AML enable only v4l2 statefull but disable v4l2 request stuff etc. This can either be handled by integration branches or by patchsets.

With RPi things get trickier.

First of all the patchset is huge (2MB) which isn't nice to have in the LE tree. Most of that is the QPU accelerated HEVC decode code for RPi3, which might be adapted to KMS/DRMPRIME (moving to shader code and decoupled from legacy MMAL) and if that's working we'll want that in LE, too.

Then format-patch won't work as jc's branch has non-linear history and slapping a diff into the LE tree isn't too nice.

Also we need to add v4l2 stateful drmprime patches on top of that (but not any request ones as they are already included in jc's branch) as that'll be needed for V4L2 H264 decoders (plus MPEG4/MPEG2/VC1 on RPi0-3).

Although merging stateful drmprime works without conflicts (even with patch) ATM I'd rather handle that via a git (integration) branch than with patchsets.

So, using an integration branch only for RPi would be an option, too.

Forgot to add: I've updated this PR with the ffmpeg merge script and pushed separate non-rpi and rpi integration branches generated by it
https://github.com/LibreELEC/FFmpeg/commits/4.3-libreelec
https://github.com/LibreELEC/FFmpeg/commits/4.3-libreelec-rpi

File: tools/ffmpeg/merge-feature-branches.sh [hunk 1]
```diff
@@ -2,16 +2,15 @@
...
+  git pull --no-ff --log "$@"
 }
 
-merge https://github.com/xbmc/FFmpeg.git 4.3-Matrix-Alpha1
-merge https://github.com/lrusak/FFmpeg.git v4l2-drmprime-v5
-merge https://github.com/Kwiboo/FFmpeg.git v4l2-request-hwaccel-4.3
-merge https://github.com/LibreELEC/FFmpeg.git 4.3-libreelec-misc
+pull https://github.com/xbmc/FFmpeg.git 4.3-Matrix-Alpha1
+pull https://github.com/lrusak/FFmpeg.git v4l2-drmprime-v5
+pull https://github.com/Kwiboo/FFmpeg.git v4l2-request-hwaccel-4.3
```
Comment (Kwiboo) [inline tools/ffmpeg/merge-feature-branches.sh:11]: You should skip this for `--rpi` or there may be conflicts when pulling in the `rpi` branch as they can become out-of-sync.

I will push support for 4:2:2 and 10-bit formats any day now and that will cause conflicts with changes from `rpi` branch until that branch is updated.

Ideally we can incorporate the rpi changes directly into the v4l2-request-hwaccel work but the rpi hevc threading part needs to be reworked before that can happen.

Comment (HiassofT) [reply tools/ffmpeg/merge-feature-branches.sh:11]: Yes, that can be dropped for `--rpi` - I mainly kept that in to reflect current state of branches

**Comment by HiassofT** [comment]
superseded by #4481

**Pull request closed without merging**